### PR TITLE
feat: support deployment behind a reverse proxy under a sub-path

### PR DESCRIPTION
**Pull request by Pfannkuchensack**
## Summary

The web frontend derived all URLs from `window.location.origin` and used hardcoded paths (`/api`, `/ws/socket.io`, `/locales`, `/openapi.json`), so running Invoke behind a reverse proxy under a sub-path (e.g. `https://host/invoke/`) broke because the prefix was lost.

**Frontend:** adds a shared helper (`common/util/baseUrl.ts`) that auto-detects the deployment prefix from the entry bundle URL (`import.meta.url`), relying on Vite's existing `base: './'`. `getBaseUrl()`, the socket.io path, the i18n `loadPath`, the openapi fetch and the React Router `basename` now all use it. At the domain root the prefix is empty, so behavior is byte-identical to before. Covered by unit tests.

**Backend:** adds an opt-in `base_url` setting (plus `forwarded_allow_ips`). When set, the app is wrapped in `SubPathASGIMiddleware`, which strips the prefix from the request path and advertises `root_path`, handling **both** proxy styles (preserve and strip). uvicorn's `root_path` is intentionally **not** used, because it *prepends* the prefix to the request path and breaks the preserve case (the prefix would appear twice). Default (unset) leaves existing installs completely unchanged.

Two deployment scenarios are supported:
- **Strip proxy** (proxy removes the sub-path before forwarding): works with zero backend config via frontend auto-detection. Setting `base_url` is optional but recommended so `/docs` + openapi URLs are correct.
- **Preserve proxy** (proxy forwards the sub-path unchanged): requires `base_url` so the middleware can strip the prefix for routing.

## QA Instructions

Tested end-to-end with Docker + a Caddy reverse proxy in three configurations; every endpoint was checked for the correct **content type and body** (not just status code):

| Configuration | UI | API (real JSON) | openapi.json | locales | `/docs` | WebSocket |
|---|:---:|:---:|:---:|:---:|:---:|:---:|
| Preserve proxy + `base_url=/invoke` | ✅ | ✅ | ✅ | ✅ | ✅ (refs `/invoke/openapi.json`) | ✅ (valid sid) |
| Strip proxy + `base_url=/invoke` | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
| Strip proxy, zero-config (`base_url` unset) | ✅ | ✅ | ✅ | ✅ | – | ✅ |

The frontend prefix-detection logic is also covered by unit tests (`baseUrl.test.ts`).

To reproduce locally (production build required — the dev server has no `/assets/` segment, so auto-detection is a no-op there):

**Strip proxy (Caddy):**
```
:8080 {
    handle_path /invoke/* {
        reverse_proxy 127.0.0.1:9090
    }
    redir /invoke /invoke/ 308
}
```
Leave `base_url` unset (or set `base_url: /invoke` to also fix `/docs`).

**Preserve proxy (nginx):**
```nginx
location /invoke/ {
    proxy_pass http://127.0.0.1:9090/invoke/;   # note: NO trailing-slash stripping
    proxy_http_version 1.1;
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
}
location = /invoke { return 301 /invoke/; }
```
Set `base_url: /invoke` and `forwarded_allow_ips: <proxy-ip>` in `invokeai.yaml`.

Then browse `http://localhost:8080/invoke/` and confirm: UI loads, generations stream progress (websocket), language switching loads locales, and a reload restores client state. Also verify a regression check at the domain root (`http://localhost:9090/`) still works with `base_url` unset.

## Merge Plan

Normal merge. No DB schema or redux slice changes. The config schema gains two opt-in fields (`base_url`, `forwarded_allow_ips`); both default to a no-op, so no migration is needed.

## Checklist

- [x] _The PR has a short but descriptive title, suitable for a changelog_
- [x] _Tests added / updated (if applicable)_
- [ ] _❗Changes to a redux slice have a corresponding migration_
- [x] _Documentation added / updated (if applicable)_
- [ ] _Updated `What's New` copy (if doing a release after this PR)_

### CODE REVIEW

File: invokeai/app/api_app.py [hunk 1]
```diff
@@ -143,6 +144,41 @@ async def dispatch(self, request: Request, call_next: RequestResponseEndpoint):
...
+        self.app = app
+        self.base_path = base_path
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if scope["type"] in ("http", "websocket"):
+            scope = dict(scope)
+            path: str = scope.get("path", "")
+            # Preserve-style proxy: strip the prefix from the path so routing matches.
+            if path == self.base_path or path.startswith(self.base_path + "/"):
+                scope["path"] = path[len(self.base_path) :] or "/"
```
Comment (lstein) [inline invokeai/app/api_app.py]: Stripping the prefix from `scope["path"]` is what makes server-generated redirects lose the sub-path (see review body: trailing-slash 307s and the root `?__theme=dark` redirect both emit `Location` without `/invoke`).

Per the ASGI spec, `scope["path"]` should *retain* the `root_path` prefix. Starlette already handles the rest: its router strips `root_path` for matching (`starlette._utils.get_route_path`), and it builds redirect/`url_for` URLs from the full `path`. So the rewrite can be inverted — **prepend** the prefix for strip-style requests instead of stripping it for preserve-style ones:

```python
if scope["type"] in ("http", "websocket"):
    scope = dict(scope)
    path: str = scope.get("path", "")
    # Strip-style proxy: the prefix is missing — restore it so scope["path"] is the
    # public path (per ASGI spec, path includes root_path). Starlette's router strips
    # root_path for matching and uses the full path for redirects/url_for.
    if not (path == self.base_path or path.startswith(self.base_path + "/")):
        scope["path"] = self.base_path + path
        raw_path = scope.get("raw_path")
        if raw_path is not None:
            scope["raw_path"] = self.base_path.encode("latin-1") + raw_path
    scope["root_path"] = self.base_path
```

One companion tweak: with the full public path in `scope["path"]`, `RedirectRootWithQueryStringMiddleware` needs to be root_path-aware, otherwise it stops matching at the sub-path root:

```python
root_path = request.scope.get("root_path", "")
if request.url.path in ("/", root_path + "/") and request.url.query:
    return RedirectResponse(url=root_path + "/")
```

I verified this combination against the same TestClient matrix: routing still 200s in both proxy styles, trailing-slash 307s emit `Location: .../invoke/api/v1/boards/`, and `/invoke/?__theme=dark` redirects to `/invoke/`.

If you'd rather keep the strip approach, the minimal patch is just the `RedirectRootWithQueryStringMiddleware` change above — but the trailing-slash 307s would stay wrong for non-SPA API consumers.

File: invokeai/app/services/config/config_default.py [hunk 2]
```diff
@@ -257,6 +261,20 @@ class InvokeAIAppConfig(BaseSettings):
 
     model_config = SettingsConfigDict(env_prefix="INVOKEAI_", env_ignore_empty=True)
 
+    @field_validator("base_url")
+    @classmethod
+    def validate_base_url(cls, v: Optional[str]) -> Optional[str]:
```
Comment (lstein) [inline invokeai/app/services/config/config_default.py:272]: Suggest rejecting base paths whose first segment collides with a real route prefix. With `base_url: /api` the server is completely dead in *both* proxy styles (verified with a TestClient: strip-style `/api/v1/images` → 404 because the middleware eats the real `/api` prefix; preserve-style `/api/api/v1/images` → 404 because Starlette's `get_route_path` strips `root_path` from the already-stripped path a second time). Same for `/ws`, `/static`, `/docs`, `/redoc`, `/locales`, `/assets`, `/openapi.json`.

Since the failure mode gives no hint at the cause, failing fast here would save users a lot of debugging:

```python
RESERVED_BASE_URL_PREFIXES = {"api", "ws", "static", "docs", "redoc", "openapi.json", "locales", "assets"}
...
first_segment = v.split("/")[0]
if first_segment in RESERVED_BASE_URL_PREFIXES:
    raise ValueError(f"base_url must not start with reserved path segment '/{first_segment}'")
```

File: invokeai/frontend/web/src/common/util/baseUrl.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * Exported separately from {@link getDeploymentBaseUrl} so it can be unit-tested without relying
+ * on `import.meta.url`.
+ */
+export const deriveDeploymentBaseUrl = (moduleUrl: string, fallbackOrigin: string): string => {
+  try {
+    const url = new URL(moduleUrl);
+    const assetsIdx = url.pathname.indexOf('/assets/');
+    const prefix = assetsIdx >= 0 ? url.pathname.slice(0, assetsIdx) : '';
+    return `${url.origin}${prefix}`;
```
Comment (lstein) [inline invokeai/frontend/web/src/common/util/baseUrl.ts:31]: Two robustness notes:

1. Previously everything used `window.location.origin`; this now uses the *module's* origin. If the bundle is ever served from a different origin than the page (CDN-hosted assets — plausible for `package`-mode consumers), API/websocket/locale requests would silently target the asset origin instead of the app origin. Cheap guard: if `url.origin !== window.location.origin` (when `window` is available), return the fallback origin — sub-path detection on a foreign origin is meaningless anyway.
2. Nit: `indexOf('/assets/')` breaks if the deployment prefix itself contains an `/assets/` segment (e.g. `/foo/assets/bar/`); `lastIndexOf` is marginally safer since Vite emits a flat `assets/` dir.

File: invokeai/app/run_app.py [hunk 4]
```diff
@@ -87,15 +87,33 @@ def run_app() -> None:
         # imported.
         enable_dev_reload(custom_nodes_path=app_config.custom_nodes_path)
 
+    # When running behind a reverse proxy that serves the app under a sub-path (e.g. `/invoke`),
+    # wrap the app so route matching and openapi/docs work for both proxy styles (strip & preserve).
+    # Also honor X-Forwarded-* headers. Only enabled when `base_url` is set, so default installations
+    # are unaffected.
+    from typing import Any
```
Comment (lstein) [inline invokeai/app/run_app.py]: Nit: the surrounding lazy imports exist for torch-init ordering, which doesn't apply to `typing` — this can move to the top of the file. Also `proxy_headers=True` is already uvicorn's default, so the kwargs dance only really carries `forwarded_allow_ips`.

**Review by lstein — CHANGES_REQUESTED** [review]
Thanks — this is a well-constructed PR. The `import.meta.url` auto-detection on the frontend is elegant and unit-tested, the `base_url` validator normalizes correctly (I exercised `'/invoke/'`, `'//invoke//'`, `' /x '`, `''`, `'/'` — all behave as documented), defaults are a strict no-op, and the dual proxy-style handling works for the main flows. I ran the new frontend tests (6/6 pass) and exercised the middleware against Starlette 0.48/FastAPI with a TestClient to verify behavior empirically.

Requesting changes for one functional gap plus two confirmed backend edge cases; details in the inline comments where the code is in the diff.

### 1. Login/setup flows break under a sub-path (blocking)

The PR sets the React Router `basename`, but two auth components navigate with hard-coded absolute paths that bypass it (these files aren't in the diff, so noting here rather than inline):

- `invokeai/frontend/web/src/features/auth/components/LoginPage.tsx:65` — `window.location.href = '/app'` (the deliberate full reload for multiuser state isolation)
- `invokeai/frontend/web/src/features/auth/components/AdministratorSetup.tsx:63` — `window.location.href = '/login'`

Under `https://host/invoke/`, completing login or admin setup navigates to `https://host/app` / `https://host/login` — outside the proxied prefix, typically a proxy 404. Since these are the entry points when auth is enabled, sub-path support is broken for multiuser installs. Fix is small and keeps the full-reload semantics:

```ts
window.location.href = `${getBasePath()}/app`;  // and '/login' in AdministratorSetup
```

### 2. Server-generated redirects lose the prefix (confirmed empirically)

Because the middleware strips `base_path` from `scope["path"]` while Starlette builds redirect URLs from `scope["path"]` alone, every server-generated redirect drops the prefix. TestClient results with `base_url=/invoke`:

```
GET /invoke/api/v1/images        -> 200                                        (routing ✓)
GET /invoke/api/v1/boards        -> 307 Location: http://host/api/v1/boards/   (prefix lost)
GET /invoke/?__theme=dark        -> 307 Location: /                            (prefix lost)
```

The SPA always sends canonical URLs so it won't hit the trailing-slash 307s, but direct API consumers will — and the `?__theme=dark` case is exactly what `RedirectRootWithQueryStringMiddleware` exists for, and it now lands the user at the domain root. See the inline comment on `SubPathASGIMiddleware` for a suggested fix (prepend instead of strip).

### 3. `base_url` colliding with a real route prefix silently bricks the app (confirmed)

With `base_url: /api`, both delivery styles 404: `/api/v1/images` gets eaten by the strip logic, and `/api/api/v1/images` gets double-stripped by Starlette's `get_route_path` (which removes `root_path` from `path` a second time). Same for `/ws`, `/static`, `/docs`, `/redoc`, `/locales`, `/assets`, `/openapi.json`. The failure mode is a completely dead server with no hint why — see inline comment on the validator.

### Minor

- No backend tests for the middleware or validator, and both are trivially unit-testable — given the subtle proxy-style matrix above, a small test would protect this from regressing. Happy to share the TestClient harness I used for the results above.

**Review by lstein — CHANGES_REQUESTED** [review]
See inline comments above for issues picked up during code review.

**Review by lstein — APPROVED** [review]
All review comments addressed. LGTM

**Comment by lstein** [comment]
Pushed a merge of `main` to resolve the conflict with #9358 (per-user socket event scoping), which landed after your last merge of main.

Also: nice catch on the engine.io handshake 404 — the prepend approach I suggested did introduce that, and folding `base_url` into `socketio_path` is the right fix. All review findings are addressed; approval incoming.
